certora/specs/Splits.spec

```diff
@@ -0,0 +1,418 @@
+//// ## Verification of Splits.sol
+//// 
+//// `Splits.sol` is part of the Drips protocol
+//// 
+//// The splits functionality, as explained in DripsHub.sol by the developers:
+//// The user can share collected funds with other users by using splits.
+//// When collecting, the user gives each of their splits receivers a fraction of the received funds.
+//// Funds received from splits are available for collection immediately regardless of the cycle.
+//// They aren't exempt from being split, so they too can be split when collected.
+//// Users can build chains and networks of splits between each other.
+//// Anybody can request collection of funds for any user,
+//// which can be used to enforce the flow of funds in the network of splits.
+//// 
+//// ### Assumptions and Simplifications
+//// - All the internal methods are wrapped in CVT callable versions
+//// - Dummy functions used to control the SplitsReceiver[] receivers
+//// - When testing the split weight functionality we safely assumed weight < _TOTAL_SPLITS_WEIGHT
+////   since the method _assertSplitsValid() validates them before assignment
+//// 
+//// ### Properties
+
+
+/// @notice Calculate results of splitting an amount using the current splits configuration.
+/// We verify that amount == collectableAmt + splitAmt, including that no overflow occurs
+rule correctnessOfSplitResults() {
+    env e; uint256 userId; uint128 amount; uint128 collectableAmt; uint128 splitAmt;
+
+    collectableAmt, splitAmt = splitResults(e, userId, true, amount);
+
+    assert amount >= collectableAmt;
+    assert amount >= splitAmt;
+    assert amount == collectableAmt + splitAmt;
+}
+
+
+/// @notice Splits user's received but not split yet funds among receivers.
+/// @param collectableAmt The amount made collectable for the user
+/// @param splitAmt The amount split to the user's splits receivers
+rule correctnessOfSplit() {
+    env e; uint256 userId; uint256 assetId; uint128 collectableAmt; uint128 splitAmt;
+
+    uint128 splittableBefore;   uint128 collectableBefore;
+    uint128 splittableAfter;    uint128 collectableAfter;
+    
+    splittableBefore = splittable(e, userId, assetId);
+    collectableBefore = collectable(e, userId, assetId);
+
+    collectableAmt, splitAmt = split(e, userId, assetId, true);
+
+    splittableAfter = splittable(e, userId, assetId);
+    collectableAfter = collectable(e, userId, assetId);
+
+    assert splittableBefore >= splittableAfter;
+    assert collectableBefore <= collectableAfter;
+    assert splittableBefore + collectableBefore >= splittableAfter + collectableAfter;
+}
+
+
+/// @notice Collects user's received already split funds
+/// After collection, no more collectable balance should be immediately available
+rule integrityOfCollect() {
+    env e; uint256 userId; uint256 assetId;
+    uint128 collectedAmt; uint128 collectableBefore; uint128 collectableAfter;
+
+    collectableBefore = collectable(e, userId, assetId);
```

Comment:
Is the first call to 'collectable' really needed here?

Reply:
Indeed not needed, I am removing the obsolete code. Thanks.